requests-for-comments/2023-10-31-feature-flag-environments.md

```diff
@@ -0,0 +1,72 @@
+# Request for comments: Feature Flag environments
+
+## Problem statement
+
+_Who are we building for, what are their needs (include example cases), why is this important?_
+
+Users find it difficult to work with Feature Flags across different projects. Currently, to use a flag in a different project, they must create it from scratch and manually copy the configuration.
+
+## Context
+
+_What are our competitors doing, what are the technical constraints, what are customers asking for, what does the data tell us, are there external motivations (e.g. launch week, enterprise contract)?_
+
+See the earlier [RFC Feature flag environments](https://github.com/PostHog/product-internal/blob/main/requests-for-comments/2022-12-08-feature-flag-environments.md#221115---rfc-feature-flag-environments) from @annikaschmid for more context.
+
+## Design
+
+_What are the key user experience and technical design decisions / trade-offs?_
+
+### Option a) Add the ability to copy a feature flag to another project.
+
+#### Pros
+
+-   Our users already use projects as environments and copy their flags manually. This solution would free them from manual work (prone to errors).
+-   Easiest to implement
+
+#### Cons
+
+-   Projects are semantically different from environments, which may not align with the needs of larger clients with multiple apps (projects).
+
+### Option b) Introduce environments to Feature Flags only
+
+#### Pros
+
+-   Cleaner separation between projects and environments.
+-   Ability to roll out environment to other parts of the app incrementally and as needed.
+
+#### Cons
+
+-   May introduce confusion by having environments for Feature Flags but not other features like Surveys.
+
+### Option C) Introduce environments app-wide (i.e. dashboards, insights, etc.)
+
+#### Pros
+
+-   Cleanest separation of concerns, respects the distinction between projects and environments.
+
+#### Cons
+
+-   Most complex to implement
+-   Introduces another layer of complexity where it might not be needed (e.g. dashboards).
+
+Based on Annika's RFC and some preliminary discussion, I recommend implementing **Option a) Add the ability to copy a feature flag to another project.** This is the quickest win which addresses the users' requirements from the [RFC user feedback](https://github.com/PostHog/product-internal/blob/main/requests-for-comments/2022-12-08-feature-flag-environments.md#user-feedback).
+
+## Implementation
+
+In the Feature Flag view page, users will have the option to "Sync" the feature flag to other projects. To do so, they'll click the Sync button, after which a modal pops up. They can choose the project to sync the flag to via a dropdown.
```

Comment:
Will updating one feature flag automatically sync the other? Will this be a configuration option?

Reply:
After they're copied over

Reply:
@neilkakkar mentioned that users should be able to have a different configuration (e.g. targeting) for a staging flag than for a production one.

We can add the option to sync automatically, but I'd imagine that might complicate things for users (perhaps hide this behind "Advanced options").

Reply:
yeah I think "Sync" is the wrong word to use here, (and don't think we should build automatic sync _yet_ ). It's more like just "Copy"

Reply:
good point, updated the RFC and the mocks to avoid confusion